### PR TITLE
Implement chunk for different backends

### PR DESCRIPTION
**Pull request by Kelvinyu1117**
### Checklist

- [x] Confirmed that `run-checks all` script has been executed.
- [x] Made sure the book is up to date with changes in this PR.

### Related Issues/PRs

#970

### Changes

This PR is in progress, mainly utilizing the backend implementation for `chunk()` instead of the default, currently only `bool_chunk` was implemented and encountered issues for `burn-fusion`.

### Testing

As it is still in progress, it has not yet been tested.

### CODE REVIEW

**Comment by Kelvinyu1117** [comment]
Currently, this PR has an error of not implementing the traits:
```
error[E0046]: not all trait items implemented, missing: `bool_chunk`
  --> burn-fusion/src/ops/boolean.rs:17:1
   |
17 | impl<B: FusionBackend> BoolTensorOps<Self> for Fusion<B> {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ missing `bool_chunk` in implementation
   |
   = help: implement the missing item: `fn bool_chunk(_: <backend::Fusion<B> as burn_tensor::backend::Backend>::BoolTensorPrimitive<D>, _: usize, _: usize) -> std::vec::Vec<<backend::Fusion<B> as burn_tensor::backend::Backend>::BoolTensorPrimitive<D>> { todo!() }`
```

I think once we discuss the solution for `burn-fusion`, the implementation of other data type (Int/Float) will be trival.

**Comment by Kelvinyu1117** [comment]
The current solution is to have a default implementation on all the `ops`, such that the backend can fall back to the default.

File: burn-tensor/src/tensor/api/base.rs [hunk 1]
```diff
@@ -1159,6 +1125,100 @@ pub trait BasicOps<B: Backend>: TensorKind<B> {
...
+        dim: usize,
+        start: usize,
+        length: usize,
+    ) -> Self::Primitive<D> {
+        let ranges: Vec<_> = (0..D)
+            .map(|i| {
+                if i == dim {
+                    start..(start + length)
+                } else {
+                    0..Self::shape(&tensor).dims[i]
```
Comment (nathanielsimard) [inline burn-tensor/src/tensor/api/base.rs:1156]: I would extract the shape before the loop, just a bit cleaner imo.

**Review by nathanielsimard — CHANGES_REQUESTED** [review]
I think you can remove the default function implementation of `chunk` and `narrow` in the `Kind` API. I think you should create files in `burn_tensor` for `narrow` and `chunk` where implementations will be located, which can reduce code duplication.

File: burn-tensor/src/tensor/api/base.rs [hunk 2]
```diff
@@ -1159,6 +1125,100 @@ pub trait BasicOps<B: Backend>: TensorKind<B> {
...
+    /// or use this function directly.
+    ///
+    /// For splitting a tensor, users should prefer the [Tensor::chunk](Tensor::chunk) function,
+    /// which is more high-level and designed for public use.
+    fn chunk<const D: usize>(
+        tensor: Self::Primitive<D>,
+        chunks: usize,
+        dim: usize,
+    ) -> Vec<Self::Primitive<D>> {
+        let size = Self::shape(&tensor).dims[dim];
```
[the remark this answers is not in the document]
Comment (Kelvinyu1117) [reply burn-tensor/src/tensor/api/base.rs:1192]: Just to confirm the interfaces, I presume they should be the following right?:
```
pub(crate) fn chunk<B: Backend, const D: usize, K: TensorKind>(
    tensor: Tensor<B, D, K>,
    chunks: usize,
    dim: usize,
) -> Vec<Tensor<B, D, K>> {
}

pub(crate) fn narrow<B: Backend, const D: usize, K: TensorKind>(
    tensor: Tensor<B, D, K>,
    dim: usize,
    start: usize,
    length: usize,
) -> Tensor<B, D, K> {
}
```

[the remark this answers is not in the document]
Comment (nathanielsimard) [reply burn-tensor/src/tensor/api/base.rs:1192]: @Kelvinyu1117 Yes sorry for the delay, I wasn't available last week 😅

File: burn-tensor/src/tensor/ops/tensor.rs [hunk 3]
```diff
@@ -1075,4 +1075,88 @@ pub trait TensorOps<B: Backend> {
...
+    ///
+    /// * `tensor` - The tensor.
+    /// * `chunks` - The number of chunks to be produced
+    /// * `times` - The dimension along which the tensor will be split.
+    ///
+    /// # Returns
+    ///
+    /// A vectors of tensors
+    ///
+    fn chunk<const D: usize>(
```
Comment (louisfd) [inline burn-tensor/src/tensor/ops/tensor.rs:1127]: It seems you have started your branch before #1006 was merged. 
In #1006 we modified the default chunk algorithm --don't mind the name of the PR which is about fixing the docs, at first I thought the doc was wrong and the code was right, but in the end the doc was right and the code was wrong. 
Could you please adjust the algorithm so it passes the test in #1006 (or in main as #1006 was merged) ?

**Review by louisfd — CHANGES_REQUESTED** [review]
Thanks, it's good that we can leverage tch and candle implementations! 
It appears we have worked a bit in parallel, see my other comment.

File: burn-tensor/src/tensor/ops/tensor.rs (hunk 3 above)
Comment (Kelvinyu1117) [reply burn-tensor/src/tensor/ops/tensor.rs:1127]: Sure, let me work on that.

**Review by louisfd — APPROVED** [review]

File: burn-tensor/src/tensor/api/base.rs [hunk 4]
```diff
@@ -1420,6 +1387,14 @@ impl<B: Backend> BasicOps<B> for Bool {
...
         B::bool_cat(vectors, dim)
     }
+
+    // fn chunk<const D: usize>(
+    //     tensor: Self::Primitive<D>,
+    //     chunks: usize,
+    //     dim: usize,
+    // ) -> Vec<Self::Primitive<D>> {
+    //     B::bool_chunk(tensor, chunks, dim)
+    // }
```
Comment (nathanielsimard) [inline burn-tensor/src/tensor/api/base.rs:1397]: Deed code

File: burn-tensor/src/tensor/api/chunk.rs [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+///
+/// A vectors of tensors
+///
+/// # Remarks
+///
+/// This is a fallback solution that used only when the backend doesn't have the corresponding implementation.
+/// Ideally, it is supposed to be implemented by the backend and the backend implementation will be resolved
+/// by static dispatch. It is not designed for direct usage by users, and not recommended to import
+/// or use this function directly.
+///
```
Comment (nathanielsimard) [inline burn-tensor/src/tensor/api/chunk.rs:22]: Small details, but I think spacing between the comments and the imports would be nice. Also, the last empty line of the comment isn't necessary.

**Review by nathanielsimard — CHANGES_REQUESTED** [review]
Some smalls changes, but it's getting ready to be merged.

File: burn-tensor/src/tensor/api/base.rs (hunk 4 above)
Comment (Kelvinyu1117) [reply burn-tensor/src/tensor/api/base.rs:1397]: fixed.

File: burn-tensor/src/tensor/api/chunk.rs (hunk 5 above)
Comment (Kelvinyu1117) [reply burn-tensor/src/tensor/api/chunk.rs:22]: fixed.

**Review by nathanielsimard — APPROVED** [review]

**Pull request merged**